---
title: "Differential Expression with Bioconductor"
output:
  pdf_document: default
  html_notebook: default
---

Introduction
This tutorial and accompayning data was provided by Randy Williams. It focuses on how to perform a differential expression analysis on microarray gene expression data with Bioconductor. it will use gene expression data from a study on tobacco smoke-related transcriptome alterations in the placenta. Differential expression analysis will be performed to find out which genes in the human placenta are differentially expressed by fetal sex. The dataset is normalized with the annotation and phenotype data provided.

```{r}
#Load Bioconductor
library(BiocManager)
```
```{r}
#Required BioConductor Packages
BiocManager::install(c('EnhancedVolcano', 'Biobase', 'limma', 'geneplotter', 'enrichplot'))
```

```{r}
BiocManager::install('clusterProfiler')
```

```{r}
BiocManager::install('pheatmap')
#loading all Cran and Bioconductor Packages
library(Biobase)
library(limma)
library(RColorBrewer)
library(dplyr)
library(ggplot2)
library(geneplotter)
library(pheatmap)
library(enrichplot)
library(tidyr)
library(EnhancedVolcano)
library(clusterProfiler)
```
First, I load the normalized expression assay, the phenotype data and the feature annoation data for this dataset
```{r}
GSE27272Norm_exprs <- read.delim('C:/Users/samen/Desktop/Bioinformatics Projects/Differential Expression Analysis with Bioconductor/GSE27272Norm_exprs.txt')

GSE27272Norm_phenoData <- read.delim('C:/Users/samen/Desktop/Bioinformatics Projects/Differential Expression Analysis with Bioconductor/GSE27272Norm_phenoData.txt')

GSE27272Norm_featureData <- read.delim('C:/Users/samen/Desktop/Bioinformatics Projects/Differential Expression Analysis with Bioconductor/GSE27272Norm_featureData.txt')
```
The file "GSE27272Norm_exprs" contains the quantile normalized expression values for each sample in the study. The rows represent genes,which are labeled with an ensemble ids. The columns are denote with the sample ids. Each entry is an expression value for a gene in a given tissue sample.
```{r}
#View(head(GSE27272Norm_exprs))
head(GSE27272Norm_exprs)
```
The file "GSE27272Norm_phenoData" contains the phenotype data, which contains metadata information for the each sample in the study. Each column is a descriptive variable, which records charactersitics like sex,bmi, and smoking status. Each row index represents a particular sample.

```{r}
#View(head(GSE27272Norm_phenoData))
head(GSE27272Norm_phenoData)
```

The file "GSE27272Norm_featureData" contains the annotation data, which is used to characterized the genes in the expression data. Each row index is a particular gene that is also found in the expression data. The columns Probe IDs, Symbols, entrez IDs and ensembl IDs are all different labels to identify the features. The column "CHR" indicates the chromosome the gene is located on. "Start.Pos" and "End.Pos", shows the position where the coding sequence of a gene starts and stops.

```{r}
#View(head(GSE27272Norm_featureData))
head(GSE27272Norm_featureData)
```

Now that the data is loaded, I will create an Expression Set with the expressino assay phenotype and the feature annotation data. An ExpressionSet is a standardized data structure in Bioconductor (via the BioBase library) which combines several different sources of information conveniently into one object

```{r echo=FALSE, out.width='100%'}
knitr::include_graphics('C:/Users/samen/Desktop/Bioinformatics Projects/Differential Expression Analysis with Bioconductor/Expression_set_image.png')
```
```{r}
#Creating an ExpressionSet object with all attributes

GSE27272_Eset <- ExpressionSet(as.matrix(GSE27272Norm_exprs))

pData(GSE27272_Eset) <- GSE27272Norm_phenoData

featureData(GSE27272_Eset) <- as(GSE27272Norm_featureData, "AnnotatedDataFrame")
```

Exploratory Graph
Before applying hypothesis testing on the data we should examine exploratory graphs like PCA and heatmaps to assess our data
```{r}
#function configuring graph size
fig <- function(width, heigth){
     options(repr.plot.width = width, repr.plot.height = heigth)
}
```

I will create PCA plot examine the variation in the data by the phenotype variable of interest
```{r}
fig(12,8)
GSE27272Norm_exprs <- Biobase::exprs(GSE27272_Eset)
PCA <- prcomp(t(GSE27272Norm_exprs), scale = FALSE )

percentVar <- round(100*PCA$sdev^2/sum(PCA$sdev^2), 1)

sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2], Phenotype = Biobase::pData(GSE27272_Eset)$sex)

ggplot(dataGG, aes(PC1, PC2)) + geom_point(aes(colour = Phenotype))+ ggtitle("PCA plot of the GSE27272") + xlab(paste0("PC1, VarExp:", percentVar[1], "%")) + ylab(paste0("PC2, VarExp:", percentVar[2], "%"))+
  theme(plot.title = element_text(hjust = 0.5, size = 25, face = 'bold'),
        axis.text.x = element_text(size = 12, face = 'bold'),
        axis.text.y =element_text(size = 12, face = 'bold'),
        axis.title.x = element_text(size = 18, face = 'bold'),
        axis.title.y = element_text(size = 18, face = 'bold'),
        legend.title = element_text(size = 18, face = 'bold'),
        legend.text = element_text(size = 18)) +scale_color_manual(values = c("hotpink", "deepskyblue"))
```
Now I will be plotting a heatmap to examine the sample to sample distances and to see how well the samples cluster to sex.

```{r}
annotation_for_heatmap <- data.frame(Phenotype = Biobase:: pData(GSE27272_Eset)$sex)

row.names(annotation_for_heatmap) <- row.names(pData(GSE27272_Eset))


dists<- as.matrix(dist(t(GSE27272Norm_exprs), method = "manhattan"))

rownames(dists) <- row.names(pData(GSE27272_Eset))
hmcol <- rev(colorRampPalette(RColorBrewer::brewer.pal(9, "YlOrRd"))(255))

colnames(dists) <- NULL
diag(dists) <- NA

ann_colors <- list(Phenotype = c(female = "hotpink", male = "deepskyblue"))

pheatmap(dists, col = (hmcol), annotation_row = annotation_for_heatmap, annotation_colors = ann_colors, legend = TRUE, treeheight_row = 0, legend_breaks = c(min(dists, na.rm = TRUE), max(dists, na.rm = TRUE)), legend_labels = (c("small distance", "large distance")), main = "Clustering heatmap for the GSE2727 samples")
```
Filtering Data
Sometimes when performing a differential expression analysis we have to subset the genes we are testing based off the annotation data. For example, if we are doing a differential expression analysis by sex it would make sense to filter out the genes on the Y chromosome. Biologically, a male has a X and Y sex chromosome while a female has two X chromosomes. Features on the Y chromosome should have no expression for females because they have no Y chromosome. Therefore, we cannot compare the difference in expression between males and females for Y-linked genes.

```{r}
# Filters the ExpressionSet (which includes the feature data and the expression data) 
# to the genes that are not present in the Y chromosome 

GSE27272_noY <- GSE27272_Eset[GSE27272_Eset@featureData@data$CHR != "Y",]
```

HYPOTHESIS TESTING
Our next aim is to perform hypothesis testing on all of our genes after filtering the data. We do this by fitting a linear model for every gene and defining contrasts to test our hypotheses. In our case, our contrasts are "female" and "male" because we are interested in finding genes in the placenta are differentially expressed by the sex of the fetus. The first step is to create a design matrix for variable of interest.

```{r}
design <- model.matrix(~0+GSE27272Norm_phenoData$sex)

colnames(design) <- c("female", "male")
GSE27272_samples <- as.character(GSE27272Norm_phenoData$geo_accession)

rownames(design) <- GSE27272_samples

#levels(GSE27272Norm_phenoData$characteristics_ch1.1) <- c("non_smoker","smoker")
#design <- model.matrix(~0+GSE27272Norm_phenoData$characteristics_ch1.1)
#colnames(design) <- c("non_smoker","smoker")
#GSE27272_samples <- 
  #as.character(GSE27272Norm_phenoData$geo_accession)
#rownames(design) <- GSE27272_samples
```

Next, we create a constrast matrix that uses the function 'makeContrasts' from the 'limma' package that will take our design matrix as an input for the levels. Afterwards, we fit a linear model with function contrast.fit() from the 'limma' package in order to examine the relationship gene expression and our variable of interest. We use the function eBayes() on our linear model to get moderated t-test statistics. The eBayes() function performs the empirical Bayes method to squeeze the gene-wise residual variance towards a pooled estimate. Moderating the test-statistics with empirical Bayes method increases the statistical power of the differential expression analysis.

```{r}
contrast_matrix <- makeContrasts(female-male, levels= design)
#contrast_matrix <- makeContrasts(non_smoke-smoker, levels = design)

GSE27272_fit <- eBayes(contrasts.fit (lmFit(GSE27272_noY, design = design),
                                      contrast_matrix))
```


Next, we make a histogram showing the frequency of pvalues. Ideally, the histogram should be a right or positively skewed distribution. If not, this could be because we did not consider batch effects or failed consider possible covariates that we need to adjust for in the hypothesis testing. Futhermore, this would lead to quality loss in later analyses.
```{r}
table_GSE27272 <- topTable(GSE27272_fit, number = Inf, confint = TRUE)

hist(table_GSE27272$P.Value, col = brewer.pal(3, name = "Set2") [1],
     main = "Female vs Male - GSE27272", xlab = "p-values")
```
Below I will make a table of the results of from differential expression analysis. The columns first three columns are Ensembl_IDs,Entrez_IDs, and Symbol, which are gene labels. Next we have the columns pvalue and adj.pvalue (adjusted pvalue), which are calculated using the limma package functions. We need to have adjusted p-values because we are doing mutiple independent tests, which could result in a large number of false positives. Another column is Log2FC,which stands for Log2 fold change. The fold change measures the effect size and informs us of the direction of the causal relationship. I will talk more about fold changes in the next section called Volcano Plot. The left and right end of the confidence intervals are denoted with CI.L and CI.R . Finally, there is the t-test statistic that is respresented with the column t
```{r}
GSE27272_Results <- data.frame(Ensembl_IDs = table_GSE27272$Ensembl_IDs,
                               Entrez_IDs = table_GSE27272$Entrez_IDs,
                               Symbol = table_GSE27272$Symbols,
                               Log2FC = table_GSE27272$logFC,
                               pvalue = table_GSE27272$P.Value,
                               adj.pvalue = table_GSE27272$adj.P.Val,
                               CI.R = table_GSE27272$CI.R,
                               t = table_GSE27272$t,
                               stringsAsFactors = FALSE)

head(GSE27272_Results)
```
Volcano Plots

After performing hypothesis testing, it is ideal for us to try visual our results. The most common graph made are volcano plots, which are scatter plots of the fold change versus the p-value for each gene. Pvalues are usually transformed on the -log10 scale. This means the more significant the p-value is (or the smaller the p-value is), the larger the value for the -log10 pvalue. A -log10 pvalue greater than -log10(.05) is statistically significant.
```{r}
volcano_names <- ifelse(abs(GSE27272_fit$coefficients) >= 1,
                        as.character(GSE27272_fit$genes$Symbols),
                        NA)

volcanoplot(GSE27272_fit, coef = 1L, style = "p-value", highlight = 100, 
            names = volcano_names,
            xlab = "Log2 Fold Change",
            ylab = NULL,
            pch = 16,
            cex = 0.35)
```
The graph above is an output of the basic volcano plot function from the limma package. This a good plot to do if you want a quick visualization of the differentially expressed genes. However, if you would want a graph that looks worthy of putting in a publication,then 'EnhancedVolcano' library should be used. The code and output of a graph from EnhancedVolcano is shown below.

```{r}
EnhancedVolcano(GSE27272_Results, 
                lab = as.character(table_GSE27272$Symbols),
                x = 'Log2FC',
                title = "GSE27272 Volcano Plot Female Vs Male",
                y = 'pvalue')
```
A gene is statistically significant after mutliple testing correction if it has an adjusted p value less than .05 . The log2 fold change determines if a gene has a biologically significant change in expression (or a gene having a meaningful effect size) between both groups. For this example, Log2 fold change determines whether a gene is upregulated or downregulated with respect to the reference group, which is placenta of female fetuses. If log2 fold change is greater than 0, then the gene expression is higher in placenta of female fetuses compared placenta of a male fetuses. On the other hand, if the log2 fold change is less than 0, then the gene expression is lower in the placentas of female fetus compared to male. If a gene's log2 fold change is too close to zero in either direction, it is hard to claim that there is a biologically meaningful effect in the genes expression with respect to fetal sex, regardless of the statistical significance of the gene. Therefore, it's important to set a threshold to determine whether gene has a biological signficant effect size. For our purposes, we are setting the criteria to be a log2 fold change that has an absolute value greater than 1.

For the volcano plot produced using EnhancedVolcano, all of the genes shown in gray are non-significant both statistically and in effect size. The genes in blue have statistically significant p-values but didn't a log2 fold change that suggests biologically significance. The green labeled genes aren't statistically significant but has a biologically meaningful log2 fold change. The genes colored in red are both statistically significant and has a biologically meaningful effect size.

Gene Enrichment Analysis and Over-Representation Analysis
Next, I would like to assess if the differentially expressed genes were associate with certain biological processes by performing a functional enrichment analysis. One type of enrichment analysis for determining pathways is an Over-representation analysis. For an over-representation analysis, we find which differentially expressed genes are annotated to a specific gene set of a database,then we test whether the overlap of these genes is significantly over-respresented in the gene set with a p-value calculated from a hypergeometric distribution. The p-values are correct for multiple testing correction.

Below is the list of statistically genes after multiple testing correction that had a meaningful effect size. These are the genes which I will test for enrichment analyses.
```{r}
sigGenes <- GSE27272_Results[ GSE27272_Results$adj.pvalue < 0.05 & !is.na(GSE27272_Results$adj.pvalue) & abs(GSE27272_Results$Log2FC) >1, ]

sigGenes
```
Now we will use the function enrichKEGG that will perform an enrichment analysis using the Kyoto Encyclopedia of Genes and Genomes (KEGG) database.
```{r}
sigGenes <- GSE27272_Results$Entrez_IDs[GSE27272_Results$adj.pvalue < 0.05& 
                                          !is.na(GSE27272_Results$adj.pvalue) >1 ]

sigGenes <- na.exclude(sigGenes)

kk <- enrichKEGG(gene = sigGenes, organism = 'hsa' )

head(kk, n =10)
```
Our gene list is significant after mutiple testing correction on the Wnt signaling pathway for the genes PPARD and EP300. Wnt signal pathway is a group of signal transduction pathways with proteins that pass signals into a cell through cell receptor surfaces. Another significant pathway after multiple testing correction is Prostate Cancer for genes ETV5 and PPARD.















